### PR TITLE
Drag and drop multiple images onto the canvas

### PR DESCRIPTION
**Pull request by bkrmendy**
## Problem:
When multiple images are dropped on the canvas, only one is inserted.

AND

mocked UIDs are not automatically reset after each test, leading to previously unnoticed flakiness.

## Fix:
Implements inserting multiple images via the strategies, and adds a mocha "hook" to reset mocked UIDs after each test.

## Details:
- adds a new function (`generateUidWithExistingComponentsAndExtraUids`) for generating `uid`s, where the generation process takes into account the `uid`s existing in the project and an array of caller-provided `uid`s (that might not be part of the project yet)
- adds a new `InteractionSubject`, which consists of multiple elements
- reset mocked UIDs after each test
- tests

### CODE REVIEW

File: editor/src/components/canvas/canvas-strategies/canvas-strategies.tsx [hunk 1]
```diff
@@ -119,11 +120,27 @@ export function pickCanvasStateFromEditorStateWithMetadata(
   }
 }
 
+function insertionSubjectsFromInsertMode(subject: InsertionSubject): InsertionSubjects {
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/canvas-strategies.tsx:123]: nit: shouldn't this be `getUnsertionSubjectsFromInsertMode`?

File: editor/src/templates/image-dnd.spec.browser2.tsx [hunk 2]
```diff
@@ -353,7 +358,107 @@ export var storyboard = (
...
+
+    fireEvent(canvasControlsLayer, makeDragEvent('drag', canvasControlsLayer, endPoint, files))
+
+    fireEvent(canvasControlsLayer, makeDragEvent('drop', canvasControlsLayer, endPoint, files))
+
+    await editor.getDispatchFollowUpActionsFinished()
+
+    await wait(250) // read the image
+
+    FOR_TESTS_CLEAR_MOCK_NEXT_GENERATED_UIDS()
```
Comment (balazsbajorics) [inline editor/src/templates/image-dnd.spec.browser2.tsx:402]: is this still needed here?

File: editor/src/templates/image-dnd.spec.browser2.tsx [hunk 3]
```diff
@@ -353,7 +358,107 @@ export var storyboard = (
...
+  it('dragging multiple images from the "finder" works', async () => {
+    FOR_TESTS_setNextGeneratedUids(['1', '2', '3'])
+
+    const editor = await renderTestEditorWithProjectContent(contents, 'await-first-dom-report')
+    const canvasControlsLayer = editor.renderedDOM.getByTestId(CanvasControlsContainerID)
+
+    const files = [
+      await makeImageFile(imgBase64, 'chucknorris.png'),
+      await makeImageFile(imgBase64, 'budspencer.png'),
+      await makeImageFile(imgBase64, 'brucelee.png'),
```
Comment (balazsbajorics) [inline editor/src/templates/image-dnd.spec.browser2.tsx:378]: hahaha

File: editor/src/templates/editor-canvas.tsx [hunk 4]
```diff
@@ -1034,54 +1040,36 @@ export class EditorCanvas extends React.Component<EditorCanvasProps> {
...
-              })
+              let actions: Array<EditorAction> = []
+              let uidsSoFar: Array<string> = []
+              let subjects: Array<ElementInsertionSubject> = []
+              for (const image of images) {
+                const { actions: actionsForImage, singleSubject } = actionsForDroppedImage(image, {
+                  generateUid: () =>
+                    generateUidWithExistingComponentsAndExtraUids(
+                      this.props.editor.projectContents,
+                      uidsSoFar,
```
Comment (balazsbajorics) [inline editor/src/templates/editor-canvas.tsx:1052]: wouldn't it be better to move this whole for loop into a dedicated function, and just use the return values here?

File: editor/src/components/canvas/canvas-strategies/canvas-strategies.tsx (hunk 1 above)
Comment (bkrmendy) [reply editor/src/components/canvas/canvas-strategies/canvas-strategies.tsx]: renamed it

File: editor/src/templates/image-dnd.spec.browser2.tsx (hunk 2 above)
Comment (bkrmendy) [reply editor/src/templates/image-dnd.spec.browser2.tsx]: yes, they are automatic, removed it

File: editor/src/templates/image-dnd.spec.browser2.tsx [hunk 5]
```diff
@@ -353,7 +355,105 @@ export var storyboard = (
...
+      makeDragEvent('drag', canvasControlsLayer, { x: 5, y: 5 }, files),
+    )
+
+    fireEvent(canvasControlsLayer, makeDragEvent('drag', canvasControlsLayer, endPoint, files))
+
+    fireEvent(canvasControlsLayer, makeDragEvent('drop', canvasControlsLayer, endPoint, files))
+
+    await editor.getDispatchFollowUpActionsFinished()
+
+    await wait(250) // read the image
```
Comment (balazsbajorics) [inline editor/src/templates/image-dnd.spec.browser2.tsx:397]: is there a way to actually await some sort of event that lets us know the image is ready?

File: editor/src/templates/editor-canvas.tsx (hunk 4 above)
Comment (bkrmendy) [reply editor/src/templates/editor-canvas.tsx:1052]: move to a dedicated function

File: editor/src/templates/image-dnd.spec.browser2.tsx (hunk 5 above)
Comment (bkrmendy) [reply editor/src/templates/image-dnd.spec.browser2.tsx:397]: so the problem is that while action dispatch is synchronous, the `SAVE_ASSET` action fires off a bunch of async actions (such as reading image data and dimensions), and those are necessary for the action to really take effect. I'll check whether there's a "low hanging fruit" solution for this

File: editor/src/components/editor/editor-modes.ts [hunk 6]
```diff
@@ -16,6 +16,11 @@ export interface ElementInsertionSubject {
   parent: InsertionParent
 }
 
+export interface ElementInsertionSubjects {
+  type: 'Elements'
+  elements: Array<ElementInsertionSubject>
+}
+
```
[the remark this answers is not in the document]
Comment (bkrmendy) [reply editor/src/components/editor/editor-modes.ts:23]: I tried it and it's somewhat involved, let's move it to a different PR

**Review by seanparsons — APPROVED** [review]

File: editor/src/components/canvas/canvas-strategies/flex-resize-basic-strategy.spec.browser2.tsx [hunk 7]
```diff
@@ -45,205 +45,314 @@ async function dragResizeControl(
...
-  })
-  it('resizes a flex element from edgePosition 0.5, 1 with drag vector (-15, 25)', async () => {
-    await resizeTestColumn(edgePosition(0.5, 1), canvasPoint({ x: -15, y: 25 }), 80, 215)
-  })
-  it('resizes a flex element from edgePosition 0.5, 1 with drag vector (-15, -25)', async () => {
-    await resizeTestColumn(edgePosition(0.5, 1), canvasPoint({ x: -15, y: -25 }), 80, 165)
+
+  describe('when the element has missing dimensions', () => {
+    describe('both missing', () => {
+      describe('horizontal movement', () => {
```
[the remark this answers is not in the document]
Comment (Rheeseyb) [reply editor/src/components/canvas/canvas-strategies/flex-resize-basic-strategy.spec.browser2.tsx:253]: I didn't even know this was an option!

**Review by Rheeseyb — APPROVED** [review]

**Pull request merged**